WorkflowToken api changes

sagarkapare opened the pull request:
Initial API changes for the WorkflowToken. Opening PR to get early feedback.

sagarkapare commented:
@anew @chtyim Please help review the PR. Thanks!

anew commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java:
```diff
@@ -20,9 +20,47 @@
+   * @param key   the key representing the entry
+   * @param value the value for the key
+   */
+  void putValue(String key, String value);
```
I would call this setValue, because below you have a method getLastSetter(). Or call that one getLastPutter()? (that sounds strange)

sagarkapare replied:
Renamed to `setValue`.

anew commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java:
```diff
@@ -20,9 +20,47 @@
+   * @return the name of the node that most recently set the value for the key
+   */
+  @Nullable
+  String getLastSetterNode(String key);
```
Call this getLastSetter()?

sagarkapare replied:
Renamed to `getLastSetter`

anew commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/TokenValueWithTimestamp.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Class representing the value for the key along with the timestamp inside the {@link WorkflowToken}.
+ */
+public final class TokenValueWithTimestamp {
```
It is not the value of the token (token holds many key/values...). I think it is fine to name this just ValueWithTime

sagarkapare replied:
Renamed to `ValueWithTime`.

sagarkapare commented:
@anew Addressed comments. Please take a look again. Thanks!

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java:
```diff
@@ -20,9 +20,47 @@
+   * @param key the key to be searched
+   * @return the map of node name to the value
+   */
+  @Nullable
```
Should it be returning an empty map instead?

sagarkapare replied:
Yes. Empty map makes client code easier. updated to return empty map.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -28,6 +31,54 @@
+    if (tokenValue == null) {
+      tokenValue = new WorkflowTokenValue();
+    }
+    tokenValue.putValue(nodeName, value);
```
so you don't need to put the `tokenValue` back to the `keyTokenValue` map?

sagarkapare replied:
Fixed.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -28,6 +31,54 @@
 public class BasicWorkflowToken implements WorkflowToken {
   private Map<String, Map<String, Long>> mapReduceCounters;
 
+  private Map<String, WorkflowTokenValue> keyTokenValue = Maps.newHashMap();
```
better call it `tokenValues` or `tokenValueMap`.

sagarkapare replied:
Renamed to `tokenValueMap`

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowTokenValue.java:
```diff
@@ -0,0 +1,57 @@
+
+  private final Map<String, ValueWithTime> nodeTokenValue = Maps.newHashMap();
+
+  private String lastSetterNode = null;
```
No need to assign to null.

sagarkapare replied:
Removed the null assignment.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowTokenValue.java:
```diff
@@ -0,0 +1,57 @@
+    return lastSetterNode;
+  }
+
+  @Nullable
```
How to distinguish the case where a key is missing vs key is there but with `null` value?

sagarkapare replied:
We can add another version of the `getValue` method as `getValueIfExist`. However not sure what will be the use case to distinguish between the missing key and `null` value.

sagarkapare replied:
Added new method `containsKey` to the `WorkflowToken` API to test the presence of the key in the `WorkflowToken`

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowTokenValue.java:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  Map<String, ValueWithTime> getAllValues() {
+    return nodeTokenValue;
```
Shouldn't return the underlying map directly, as it can get modified outside of this class.

sagarkapare replied:
Returning immutable copy of the map now.

anew commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java:
```diff
@@ -20,9 +20,47 @@
+
+  /**
+   * Put the specified key-value entry in the {@link WorkflowToken}.
+   * The value is stored against the node at which it is being added.
```
The token may store additional information about the context in which this key is being set, for example, the unique name of the workflow node.

sagarkapare replied:
Fixed the documentation.

sagarkapare commented:
@chtyim @anew Addressed review comments. Please take a look again. Thanks!

sagarkapare commented:
Design document for the WorkflowToken API -https://wiki.cask.co/display/CE/WorkflowToken+Enhancements

Jira: https://issues.cask.co/browse/CDAP-2709

sagarkapare commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java:
```diff
@@ -16,18 +16,113 @@
+   * @return the list of {@link NodeValueEntry} from node name to the value that node
+   * added for the input key
+   */
+  List<NodeValueEntry> getAll(String key);
```
For now keeping the return type as `NodeValueEntry` instead of `Map<Entry, Entry>`.  `NodeValueEntry` has an advantage as by calling `NodeValueEntry.getNodeName()` we are making it explicit that the `nodeName` would be return, however in case of `Map<Entry, Entry>` in order to get `nodeName`, we will need to call `entry.getKey()`.  Using `Map<Entry, Entry>` may look more natural in the API. No strong opinion. Please let me know which one should be used.

chtyim replied:
What is `Map<Entry, Entry>`? Do you mean `List<Map.Entry<String, String>>`? Or something else?

sagarkapare replied:
Yes. Keeping the return type of `List<NodeValueEntry>` instead of `List<Map.Entry<String, String>`.

sagarkapare commented:
@anew @chtyim Please help review.

anwar6953 commented on cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceContext.java:
```diff
@@ -123,4 +126,12 @@
+  /**
+   * @return the {@link WorkflowToken} associated with the current {@link Workflow},
+   * if the {@link MapReduce} program is executed as a part of the Workflow.
+   *
```
extra newline

sagarkapare replied:
Removed.

anwar6953 commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java:
```diff
@@ -16,18 +16,113 @@
+  /**
+   * This method is deprecated as of release 3.1. Instead to get the
+   * MapReduce counters from the WorkflowToken, use the flatten key prefixed
+   * by 'mr.counters.'.
```
Can you build the javadocs and make sure its properly formatted?
`mvn clean package javadoc:javadoc -pl cdap-api -am -DskipTests -P release`

sagarkapare replied:
Updated the javadoc and verified the formatting.

anwar6953 commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java:
```diff
@@ -80,6 +82,7 @@ public BasicMapReduceContext(Program program,
                                MapReduceSpecification spec,
                                long logicalStartTime,
                                String workflowBatch,
+                               WorkflowToken workflowToken,
```
mark as `@Nullable`?

anwar6953 commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/DynamicMapReduceContext.java:
```diff
@@ -74,6 +75,7 @@ public DynamicMapReduceContext(Program program,
                                  Arguments runtimeArguments,
                                  MapReduceSpecification spec,
                                  long logicalStartTime, String workflowBatch,
+                                 WorkflowToken workflowToken,
```
mark as `@Nullable`?

anwar6953 replied:
Hmm, other parameters of this constructor aren't appropriately marked. Maybe the inconsistency would be more confusing. Unsure.

sagarkapare replied:
Marked them as `@Nullable`

bdmogal commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/NodeValueEntry.java:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Can we start using Java 7 ``Objects`` class for all ``equals``, ``hashCode`` and ``toString`` now?

sagarkapare replied:
Not sure how the `Objects` class would be useful here for the class which only contain primitive types.
For example, current `equals` method return `nodeName.equals(that.nodeName) && value.equals(that.value)`. Changing it to `Objects.equals(this.nodeName, that.nodeName) && Objects.equals(this.value, that.value)` would be more verbose.. no?

bdmogal replied:
Maybe not for the equals method. How about ``hashCode`` and ``toString``?

sagarkapare replied:
Since the class `NodeValueEntry` itself only contains 2 members of type `String`, not sure how the `Objects.toString` would be better either. 

For example, current `toString` directly uses the `nodeName` and `value` instances. We should not use `Objects.toString(nodeName)` and `Objects.toString(value)` right? 

Had `nodeName` and `value` been the instances of another object, then yes you are right we would require `Objects.toString`.

Same with the `hashCode` method.

Please let me know if I am missing anything here.

bdmogal replied:
I see. ignore.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/AbstractMapReduceContextBuilder.java:
```diff
@@ -109,8 +112,8 @@ public BasicMapReduceContext build(MapReduceMetrics.TaskType type,
       new BasicMapReduceContext(program, type, RunIds.fromString(runId), taskId, runtimeArguments, datasets, spec,
-                                logicalStartTime, workflowBatch, discoveryServiceClient, metricsCollectionService,
-                                datasetFramework, adapterSpec, pluginInstantiator);
+                                logicalStartTime, workflowBatch, workflowToken, discoveryServiceClient,
```
Does ``WorkflowContext``or anything similar contain both ``WorkflowToken`` and the name of the workflow? If so, can we use that here instead of two distinct parameters representing a workflow?

sagarkapare replied:
`workflowBatch` carries the name of the current program which Workflow wants to run, which is not included in the `WorkflowContext`.

chtyim replied:
The name of the current program? So it's not the unique name (node name)?

sagarkapare replied:
`workflowBatch` is the part of existing code. We pass the name of the program in it, so that we can correctly initialize the `Program` from its specifications. It won't be the unique name of the node.

chtyim replied:
So what is this name? Do you have an example?

sagarkapare replied:
So for example when we run the `PurchaseHistoryWorkflow`, workflowBatch will be the `PurchaseHistoryBuilder`, the MapReduce program run by the Workflow. In `AbstractMapReduceContextBuilder`, while building the MapReduce context, if workflowBatch is not null, then we create the `Program` by fetching the specification for the MapReduce program from application specifications.
Corresponding code snippet in `AbstractMapReduceContextBuilder`:
```java
    // Initializing Program
    Program program = mrProgram;

    // See if it was launched from Workflow; if it was, change the Program.
    if (workflowBatch != null) {
      MapReduceSpecification mapReduceSpec = program.getApplicationSpecification().getMapReduce().get(workflowBatch);
      Preconditions.checkArgument(mapReduceSpec != null, "Cannot find MapReduceSpecification for %s", workflowBatch);
      program = new WorkflowMapReduceProgram(program, mapReduceSpec);
    }
```

chtyim replied:
So it's the program Id (the string part) of the MapReduce program, right? The variable name seems really bad.

sagarkapare replied:
yes. will change the name to something similar to `programNameInWorkflow`?

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/AbstractMapReduceContextBuilder.java:
```diff
@@ -65,6 +67,7 @@ public BasicMapReduceContext build(MapReduceMetrics.TaskType type,
                                      String taskId,
                                      long logicalStartTime,
                                      String workflowBatch,
```
I think ``workflowBatch`` should also be ``@Nullable``. Although please see https://github.com/caskdata/cdap/pull/2945/files#r33312959

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java:
```diff
@@ -122,6 +126,20 @@ public String getWorkflowBatch() {
+    }
+  }
+
+  public WorkflowToken getWorkflowToken() {
```
``@Nullable``

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java:
```diff
@@ -137,6 +139,13 @@ public ProgramController run(final Program program, ProgramOptions options) {
 
     String workflowBatch = arguments.getOption(ProgramOptionConstants.WORKFLOW_BATCH);
+
+    WorkflowToken workflowToken = null;
```
need to initialize to ``null``?

sagarkapare replied:
We need to initialize the local variables.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/AbstractProgramWorkflowRunner.java:
```diff
@@ -54,13 +55,16 @@
+  private final Arguments userArguments;
+  private final Arguments systemArguments;
+  private final String nodeId;
+
```
remove empty line

sagarkapare replied:
Removed.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -28,18 +34,137 @@
   private Map<String, Map<String, Long>> mapReduceCounters;
 
+  private Map<String, List<NodeValueEntry>> tokenValueMap = Maps.newHashMap();
+
```
remove newlines between fields

sagarkapare replied:
Removed.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -28,18 +34,137 @@
+
+  @Override
+  public void put(String key, String value) {
+    Preconditions.checkNotNull(nodeName, "Node name cannot be null.");
```
This throws ``NullPointerException`` if the precondition is not satisfied. Is that ok? Or do you think it should be an ``IllegalArgumentException``?

sagarkapare replied:
Good point! I think `IllegalStateException` would be correct to throw here. Fixed.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -28,18 +34,137 @@
+    for (Map.Entry<String, List<NodeValueEntry>> entry : otherTokenValueMap.entrySet()) {
+      if (!tokenValueMap.containsKey(entry.getKey())) {
+        // Key is newly added to the other WorkflowToken
+        tokenValueMap.put(entry.getKey(), Lists.<NodeValueEntry>newArrayList());
```
hmm, am I missing something? shouldn't this be ``entry.getValue()`` instead of empty arraylist?

sagarkapare replied:
All the values from the `entry.getValue()` will get added to the `tokenValueMap` in the for loop following on line 58.

bdmogal replied:
correct, although it might be better to use ``entry.getValue()`` here followed by ``continue``? This logic is slightly complex. Trying to see how we can remove some complexity.

sagarkapare replied:
ohh ok. yeah we can do that. updated the code.

chtyim replied:
Regarding my previous comment about immutable list, currently you store the list as ArrayList (hence mutable). My question was, would it be better to make it as an ImmutableList / UnmodifiableList (the logic of constructing the map would be slight different)? I feel more comfortable with immutable collections and I don't see the need of mutability when construction is done.

sagarkapare commented:
@bdmogal @anwar6953 Thanks for the review comments. Updated the code per feedback. Please take a look again.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java:
```diff
@@ -16,18 +16,136 @@
+   * @param prefix optional prefix to filter the keys
+   * @return the map of key-values that were added by the specified node
+   */
+  Map<String, String> getAllFromNode(String nodeName, @Nullable String prefix);
```
Is the prefix a string prefix or dot terminated prefix? For example, if prefix = "a.b", does it match with "a.b*" or only match with "a.b.*"?

sagarkapare replied:
Its a string prefix. so if prefix = "a.b", it will match with "a.b", "a.b." and "a.bc" etc. We find the key matching prefix if `key.startsWith(prefix)` is true.

anew replied:
I like Terence's idea of saying it is along dot boundaries. So prefix a.b will match a.b and a.b.c, but not a.bc

sagarkapare replied:
Sounds good! Updated the code. 
So Prefix "a.b" will match with the key "a.b" or any key starting with the "a.b.", however it will not match with the key "a.bc". To retrieve the MapReduce counters from the particular node use the prefix "mr.counters".

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -27,19 +32,139 @@
  */
 public class BasicWorkflowToken implements WorkflowToken {
   private Map<String, Map<String, Long>> mapReduceCounters;
```
Why need to keep it separate? Can it be just a filtered map view on the `tokenValueMap` for the `getMapReduceCounters` method?

sagarkapare replied:
Since the `tokenValueMap` maintains flatten counter names as `mr.counters.groupname.countername`, we won't be able to reconstruct the group name and counter name from it. I thought of maintaining the map of all counter group names and counter names and then use them to recreate the `mapReduceCounters`, however it will not work if the user has added the custom counter.

chtyim replied:
but the mapping is known to us when the raw MapReduce counters are provided this class right? However, it's not a very big deal to keep it.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -27,19 +32,139 @@
  */
 public class BasicWorkflowToken implements WorkflowToken {
   private Map<String, Map<String, Long>> mapReduceCounters;
+  private Map<String, List<NodeValueEntry>> tokenValueMap = Maps.newHashMap();
```
Can this be final? Instead of the setter method, you can have a replace method that will clear the map and replace the content.

chtyim replied:
Another question not related to the field signature. I think it's better for the actual instance of the List be immutable.

sagarkapare replied:
Made the `tokenValueMap` final. 
Can you please explain the comment 
 >> I think it's better for the actual instance of the List be immutable.
Since we keep adding the `NodeValueEntry` to the list if the `put` on the same key is made by subsequent nodes, not sure how making the List immutable will work.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -27,19 +32,139 @@
+    for (Map.Entry<String, List<NodeValueEntry>> entry : otherTokenValueMap.entrySet()) {
+      if (!tokenValueMap.containsKey(entry.getKey())) {
+        // Key is newly added to the other WorkflowToken
+        tokenValueMap.put(entry.getKey(), entry.getValue());
```
The other list may not be immutable. It's not safe to just use the List returned by getValue. Also, after the merge, you may append new entries to the List right? Meaning it will modify the List inside another object?

sagarkapare replied:
Yes. Agree. Changed the code to create the new list and adding the elements to it explicitly on the line 58

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -27,19 +32,139 @@
+    if (containsKey(key)) {
+      return ImmutableList.copyOf(tokenValueMap.get(key));
+    }
+    return Lists.newArrayList();
```
Better be `ImmutableList.of();`

sagarkapare replied:
Fixed.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -27,19 +32,139 @@
+      return null;
+    }
+    List<NodeValueEntry> nodeValueList = tokenValueMap.get(key);
+    if (nodeValueList.size() == 0) {
```
Use `isEmpty()` instead.

sagarkapare replied:
Fixed.

sagarkapare commented:
@chtyim @bdmogal Addressed comments. Please take a look again. Thanks!

anew commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java:
```diff
@@ -16,18 +16,136 @@
 public interface WorkflowToken {
+
+  /**
+   * Put the specified key-value entry in the {@link WorkflowToken}.
```
in/into

sagarkapare replied:
Changed.

anew commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java:
```diff
@@ -16,18 +16,136 @@
+   * The list maintains the order in which the values were
+   * inserted in the WorkflowToken for a specific key except in the case of fork
+   * and join. In case of fork in the Workflow, copies of the WorkflowToken are made
+   * and passed it along the each branch. At the join, all copies of the
```
remove "it"

anew replied:
I would not go into that much detail, as it limits us as to how we can change this in the future. All I would say here is that order is guaranteed for nodes within the same branch, but not across different branches.

sagarkapare replied:
Yes. I agree. Removed the extra information.

anew commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java:
```diff
@@ -16,18 +16,136 @@
+   * @param nodeName the name of the node
+   * @return the value set for the key by nodeName
+   */
+  @Nullable
```
How about moving this up a little, to come right after put().

sagarkapare replied:
Moved up.

anew commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java:
```diff
@@ -16,18 +16,136 @@
+   * by specific node.
+   * <p>
+   * This method also accepts the optional prefix parameter. When
+   * supplied, the returned map would be filtered by the keys prefixed by the input prefix.
```
would be/is

sagarkapare replied:
Updated.

sagarkapare commented:
@anew Addressed the comments. Please take a look again. Thanks!

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -27,19 +32,138 @@
+      nodeValueList = Lists.newArrayList();
+      tokenValueMap.put(key, nodeValueList);
+    }
+    nodeValueList.add(new NodeValueEntry(nodeName, value));
```
So with this logic, it means a node can set with the same key twice with different values and both will be retained, but then the `getAllFromNode` method will only return the latest value, right? I don't find it stated clearly about this behavior in the javadoc.

sagarkapare replied:
oh ok. I think while inserting into the List, if the entry for the nodeName already exists for the specific key, that entry should be replaced by the new one. So that for any key, we will always have only one entry for the specific node. I will update the code accordingly.

sagarkapare replied:
done.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -27,19 +32,138 @@
+  @Nullable
+  @Override
+  public String get(String key) {
+    if (!containsKey(key)) {
```
you actually don't need this check. Just do a get from the map and it returns `null`, then return `null`.

sagarkapare replied:
Fixed.

sagarkapare commented:
@chtyim @bdmogal Please take a look again. 
Fixed following things - 
1. Renamed the variable `workflowBatch` to the `programNameInWorkflow`.
2. Created the constructor for the `BasicWorkflowToken` which accepts the individual maps.
3. Removed prefix from the api and filed jira - https://issues.cask.co/browse/CDAP-2894

New build is running here - http://builds.cask.co/browse/CDAP-DUT2031-6

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java:
```diff
@@ -132,8 +136,17 @@ public long getLogicalStartTime() {
-  public String getWorkflowBatch() {
-    return workflowBatch;
+  public String getProgramNameInWorkflow() {
+    return programNameInWorkflow;
```
is this ``@Nullable``? If so, mark both the constructor param and this method.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java:
```diff
@@ -27,27 +32,165 @@
+            break;
+          }
+        }
+        if (!otherNodeValueEntryExist) {
```
will this condition ever be ``false``? Please check. Else remove.

bdmogal commented on cdap-app-fabric/src/test/java/co/cask/cdap/ConditionalWorkflowApp.java:
```diff
@@ -116,6 +128,31 @@ public void beforeSubmit(MapReduceContext context) throws Exception {
+      }
+
+      WorkflowToken workflowToken = context.getWorkflowToken();
+      if (workflowToken == null) {
```
nit: this is kind of ugly. apps should ideally not have to do this check. But unsure how we can fix it (if we can). Maybe lets report a jira?

bdmogal commented:
couple of minor nits. Other than that, LGTM. Presume that you have spoken with @chtyim about a couple of his comments that seem unresolved.

sagarkapare commented:
Build passed here - http://builds.cask.co/browse/CDAP-DUT2031-8. Merging.
